plugins/outputs/cloudwatchlogs/pusher.go

```diff
@@ -297,38 +308,103 @@ func (p *pusher) send() {
+			p.Log.Errorf("All %v retries to %v/%v failed for PutLogEvents, request dropped.", retryCountShort+retryCountLong-1, p.Group, p.Stream)
 			p.reset()
 			return
 		}
 
-		p.Log.Warnf("Retried %v time, going to sleep %v before retrying.", retryCount, wait)
+		p.Log.Warnf("Retried %v time, going to sleep %v before retrying.", retryCountShort+retryCountLong-1, wait)
 
 		select {
 		case <-p.stop:
-			p.Log.Errorf("Stop requested after %v retries to %v/%v failed for PutLogEvents, request dropped.", retryCount, p.Group, p.Stream)
+			p.Log.Errorf("Stop requested after %v retries to %v/%v failed for PutLogEvents, request dropped.", retryCountShort+retryCountLong-1, p.Group, p.Stream)
 			p.reset()
 			return
 		case <-time.After(wait):
 		}
 
-		retryCount++
 	}
 
 }
 
-func retryWait(n int) time.Duration {
+// retryWaitShort returns a duration to wait before retrying a request.
+// This implements an exponential backoff strategy with jitter.
+// The base wait time is 200 milliseconds, and the maximum wait time is 1 minute (jittered).
+// The wait time is capped at 1 minute if the retry count exceeds 5.
+func retryWaitShort(retryCount int) time.Duration {
 	const base = 200 * time.Millisecond
 	// Max wait time is 1 minute (jittered)
 	d := 1 * time.Minute
-	if n < 5 {
-		d = base * time.Duration(1<<int64(n))
+	if retryCount < 5 {
+		d = base * time.Duration(1<<int64(retryCount))
+	}
+	return time.Duration(seededRand.Int63n(int64(d/2)) + int64(d/2))
+}
+
+// retryWaitLong returns a duration to wait before retrying a request.
+// This implements an exponential backoff strategy with jitter.
+// The base wait time is 2 seconds, and the maximum wait time is 1 minute (jittered).
+// The wait time is capped at 1 minute if the long-retry count exceeds 2.
+func retryWaitLong(retryCount int) time.Duration {
+	const base = 2 * time.Second
+	// Max wait time is 1 minute (jittered)
+	d := 1 * time.Minute
+	if retryCount < 2 {
+		d = base * time.Duration(1<<int64(retryCount))
 	}
 	return time.Duration(seededRand.Int63n(int64(d/2)) + int64(d/2))
 }
 
+// chooseRetryWaitStrategy decides if a "long" or "short" retry strategy should be used when the PutLogEvents API call
+// returns an error. A short retry strategy should be used for most errors, while a long retry strategy is used for
+// errors where retrying too quickly could cause excessive strain on the backend servers.
+//
+// Specifically, use the long retry strategy for the following PutLogEvent errors:
+//   - 500 (InternalFailure)
+//   - 503 (ServiceUnavailable)
+//   - Connection Refused
+//   - Connection Reset by Peer
+//   - Connection Timeout
+//   - Throttling
+func chooseRetryWaitStrategy(err error) retryWaitStrategy {
+	if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+		return retryLong
+	}
+
+	// Connection reset/refused are actually syscalls
+	if errors.Is(err, syscall.ECONNRESET) || errors.Is(err, syscall.ECONNREFUSED) {
```

Comment:
Will these fall under non-AWS errors? There's this section on line 277, 
```go
awsErr, ok := err.(awserr.Error)
		if !ok {
			p.Log.Errorf("Non aws error received when sending logs to %v/%v: %v. CloudWatch agent will not retry and logs will be missing!", p.Group, p.Stream, err)
			// Messages will be discarded but done callbacks not called
			p.reset()
			return
}
```
Will this error cause the logs to get dropped?

Reply:
The way I interpret the [SDK docs](https://docs.aws.amazon.com/sdk-for-go/api/aws/awserr/#Error) is that SDK API calls should always wrap lower level errors in an awserr.Error before returning to the caller. If that's the case, then these syscall errors would be wrapped by awserr.Error, and it would be rather unexpected to ever get a non-AWS error. If that's not how it works, then yeah I believe it would cause the logs to get dropped.

Perhaps there's a way to verify that behavior (unit tests mocking the return types obviously won't work)...

Reply:
I created a unit test which replaces the underyling HttpClient used by the cloudwatchlogs client with a mock client that returns errors to see what the PutLogEvents API call returns.

```
customHTTPClient := &errorRoundTripper{
	rt:     http.DefaultTransport,
	errMsg: "connection refused",
}
			
sess, err := session.NewSession(&aws.Config{
	HTTPClient: &http.Client{
		Transport: customHTTPClient,
	},
	Region: aws.String("us-east-1"),
})
require.NoError(t, err)

cwl := cloudwatchlogs.New(sess)
params := &cloudwatchlogs.PutLogEventsInput{
	LogEvents: []*cloudwatchlogs.InputLogEvent{
		{
			Message:   aws.String("test message"),
			Timestamp: aws.Int64(time.Now().UnixNano() / int64(time.Millisecond)),
		},
	},
	LogGroupName:  aws.String("hello"),
	LogStreamName: aws.String("world"),
}
_, err = cwl.PutLogEvents(params)
fmt.Println(reflect.TypeOf(err))
require.True(t, strings.Contains(err.Error(), tt.errMsg))
_, ok := err.(awserr.Error)
require.True(t, ok, "error returned is not an awserr")
```
The unit test verifies that an error returned by the underlying HTTPClient would be wrapped by an `awserr.Error` before being returned to the caller. So, underlying errors will not get caught by the section on line 277.

However, I did find an example in aws-sdk-go where the [SDK itself checks for connection reset errors](https://github.com/aws/aws-sdk-go/blob/02c1f723b528251a45068001bee8b56d904d7484/aws/request/connection_reset_error.go) by looking at the error message instead of the error type:
```
func isErrConnectionReset(err error) bool {
	if strings.Contains(err.Error(), "read: connection reset") {
		return false
	}

	if strings.Contains(err.Error(), "use of closed network connection") ||
		strings.Contains(err.Error(), "connection reset") ||
		strings.Contains(err.Error(), "broken pipe") {
		return true
	}

	return false
}
```

So checking for the syscall errors may not be what we want to do in the agent either. It may be better to borrow the logic from the SDK since there seems to be a series of issues that lead to this particular implementation:
* https://github.com/aws/aws-sdk-go/commit/135dd4cb18a208322feca0a3f46519920beae53a
* https://github.com/aws/aws-sdk-go/commit/c3d27102088451890a9ba5be0dc1f2e0425f0fc3
* https://github.com/aws/aws-sdk-go/commit/2c957a88ad3e912c48e72b9691eba01912a0218d